java/org/apache/jasper/compiler/NewlineReductionServletWriter.java

```diff
@@ -0,0 +1,41 @@
+package org.apache.jasper.compiler;
+
+import java.io.PrintWriter;
+
+/**
+ * This class filters duplicate newlines instructions from the compiler output,
+ * and therefore from the runtime JSP. The duplicates typically happen because
+ * the compiler has multiple branches that write them, but they operate
+ * independently and don't realize that the previous output was identical.
+ *
+ * Removing these lines makes the JSP more efficient by executing fewer operations during runtime.
+ *
+ * @author Engebretson, John
+ * @author Kamnani, Jatin
+ *
+ */
+public class NewlineReductionServletWriter extends ServletWriter {
+    private static final String NEWLINE_WRITE_TEXT = "out.write('\\n');";
```

Comment:
This one is a bit strange.
I haven't written JSP code in more than 10 years now but if I want to produce several new lines I'd do `out.write("\n\n\n\n")` instead of 4 times `out.write('\\n');`.
In what cases one would have several `out.write('\\n');` ?